### PR TITLE
A PR for Issue #25 ( Continuous Integration ) 

### PR DESCRIPTION
**Pull request by JohannSuarez**
For issue #25

- Two new workflow files ( **.github/workflows/mypy.yml** and **.github/workflows/pytest.yml** )

- .**/ganache.py** now has "&" at the end for workflow to run it in the background.

- Pytest having issues with the config option "collect_ignore" in **oceanprotocol/contracts/setup.cfg**. I forked the contracts repo to remove the line. For now, the workflow clones my fork of the contracts repo. 

- In tokenspice's README, regarding the instruction about running one test,  that specified test doesn't exist anymore. ~~So I edited the README to reference an existing test, and also added the workflow status badges.~~ EDIT: BK's push seemed to have addressed this. I adjusted accordingly.

- In tokenspice.ini , ARTIFACTS_PATH now points to a user's relative directory instead of the absolute trentmc directory.

### CODE REVIEW

**Comment by trentmc** [comment]
Hi @JohannSuarez great to see this PR. Is it ready for review?

**Comment by JohannSuarez** [comment]
Hello, @trentmc! Yes, please let me know what you think and if there's something that must be changed

File: README.md [hunk 1]
```diff
@@ -177,6 +182,7 @@ source venv/bin/activate
 #run a single pytest-based test
 pytest web3engine/test/test_btoken.py::test_ERC20
 
+
```
Comment (trentmc) [inline README.md:185]: Remove this line

File: tokenspice.ini [hunk 2]
```diff
@@ -2,7 +2,7 @@
 [general]
 
 NETWORK = ganache
-ARTIFACTS_PATH = /home/trentmc/code/contracts/artifacts
+ARTIFACTS_PATH = contracts/artifacts
```
Comment (trentmc) [inline tokenspice.ini:5]: Contracts are typically _not_ going found in a subdirectory of `tokenspice` directory. They'll be found inside the contracts repo that was git-cloned. (As discussed in the call, we _will_ move it to be a subdirectory of tokenspice at some point. But that requires a bunch more work and machinery to be in place, it's outside the scope of this issue.)

Perhaps make the default `~/contracts/artifacts/` or `../contracts/artifacts`

File: README.md [hunk 3]
```diff
@@ -2,6 +2,11 @@
 
 # TokenSPICE: EVM Agent-Based Token Simulator 🐟🌪️
 
+<div align="center">
+<img alt="Pytest Unit Testing" src="https://github.com/JohannSuarez/tokenspice/actions/workflows/pytest.yml/badge.svg">
+<img alt="MyPy Static Type Checking" src="https://github.com/JohannSuarez/tokenspice/actions/workflows/mypy.yml/badge.svg">
```
Comment (trentmc) [inline README.md:7]: It tells me that the static type checking is failing. Is it passing for you? If no, what would be next steps? We don't want checks to be put in that are immediately broken.

**Review by trentmc — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (JohannSuarez) [reply README.md]: Done, thank you for pointing it out!

File: tokenspice.ini (hunk 2 above)
Comment (JohannSuarez) [reply tokenspice.ini:5]: Hello, Trent! I've tried to adjust the workflow such that the contracts repo is outside the tokenspice repo both as 
`~/contracts/artifacts/` and `../contracts/artifacts` as you have suggested. 

After several hours of debugging I discovered why the workflow was acting unpredictably. It turns out there's an inherent limitation with the [Github Action](https://github.com/actions/checkout) that is used to clone repos —  a user must clone other repos within $GITHUB_WORKSPACE ( in this case, it's the tokenspice directory. ) 

There have been several about this over the past months, see here:
https://github.com/actions/checkout/pull/388
https://github.com/actions/checkout/issues/197

I'm currently exploring possible hacks to get around this like using symlinks that point to something outside the repo root.

File: README.md (hunk 3 above)
Comment (JohannSuarez) [reply README.md:7]: Hello, Trent! The static type checks have been failing by default even when using MyPy manually outside of this Github workflow. The result of this test is true to the current status of the TokenSPICE repo. I'll have to study more about MyPy and stubs before I can completely tackle the errors.

In the meantime, we can temporarily hide the MyPy badge from the README to hide the error status, or rig it to automatically pass.

**Comment by trentmc** [comment]
Lots of progress, cool! Lmk when you would like me to review again. Best,

**Comment by JohannSuarez** [comment]
Good evening, Trent! 

I removed the whitespace, and tokenspice.ini has been modified to your requested changes.
For the failing static type checks, I will need further guidance. 

The static type checks have been failing by default even when using MyPy manually outside of this Github workflow. The result of this test is true to the current status of the TokenSPICE repo. I'll have to study more about MyPy and stubs before I can completely tackle the errors. This task could be big enough to be its own separate issue.

In the meantime, we can temporarily hide the MyPy badge from the README to hide the error status, and/or rig the static typechecking workflow to automatically pass.

**Comment by trentmc** [comment]
Re typechecks: yes let's hide it for now (as you've done) and then create a new issue for it (as you suggest).

Is the rest ready for review?

**Comment by JohannSuarez** [comment]
Okay, I have hidden the MyPy badge and set it to always pass.
Yes, I think everything is ready for review

**Comment by trentmc** [comment]
Most things look good. I run into a few errors when running mypy locally, see below. Could you fix please?

```console
(venv) trentmc@trentmcLnv:~/code/tokenspice$ mypy --config-file mypy.ini ./
assets/agents/DataconsumerAgent.py:25: error: "DataconsumerAgent" has no attribute "_buyDT"
assets/agents/DataconsumerAgent.py:26: error: "DataconsumerAgent" has no attribute "_consumeDT"
engine/test/test_SimStateBase.py:15: error: Method must have at least one argument
engine/test/test_SimEngine.py:19: error: Method must have at least one argument
assets/agents/test/test_MinterAgents.py:5: error: Incompatible import of "AgentBase" (imported name has type Module, local name has type "Type[AgentBase]")
Found 5 errors in 4 files (checked 100 source files)
```

**Review by trentmc — CHANGES_REQUESTED** [review]
Most things look good. I run into a few errors when running mypy locally, see below. Could you fix please?

```console
(venv) trentmc@trentmcLnv:~/code/tokenspice$ mypy --config-file mypy.ini ./
assets/agents/DataconsumerAgent.py:25: error: "DataconsumerAgent" has no attribute "_buyDT"
assets/agents/DataconsumerAgent.py:26: error: "DataconsumerAgent" has no attribute "_consumeDT"
engine/test/test_SimStateBase.py:15: error: Method must have at least one argument
engine/test/test_SimEngine.py:19: error: Method must have at least one argument
assets/agents/test/test_MinterAgents.py:5: error: Incompatible import of "AgentBase" (imported name has type Module, local name has type "Type[AgentBase]")
Found 5 errors in 4 files (checked 100 source files)
```

**Comment by JohannSuarez** [comment]
Hello, Trent! I have fixed all the errors that have been enumerated above.

Here's my log:
- **engine/test/test_SimStateBase.py** - A method had no arguments, so I applied the _@staticmethod_ decorator.

- **engine/test/test_SimEngine.py** -  A method had no arguments, so I also applied the _@staticmethod_ decorator.

- **assets/agents/DataconsumerAgent.py** - There was an if-block that mentioned __buyDT_ and __consumeDT_, neither of which have been defined anywhere, not even in the parent class. A ripgrep search shows these two methods are never mentioned anywhere else in the entire repo. I block commented it out, and left a message explaining why. 

- **assets/agents/test/test_MinterAgents.py** - MyPy was complaining about the line `from engine import AgentBase`, which was odd considering dozens of other files have the exact same line. I simple reworded the line to `import engine.AgentBase`, made adjustments to the lines invoking it, and the error vanished. 

These errors were actually what the MyPy workflow kept raising. Upon fixing these, the workflow now passes and so I have unhidden the MyPy badge.

**Review by trentmc — APPROVED** [review]
Looks good. Approving. Then I will merge.

Re DataconsumerAgent::_buyDT - something's up with that. So I just created #80 to fix it. The comment you inserted is appropriate for the meantime.

**Pull request merged**